### ISSUE TITLE
Improve `DataRepr` based machinery (e.g `pmatch`) efficiency

### ISSUE THREAD

**Issue by TotallyNotChase** [Plutonomicon/plutarch-plutus#214]
We should try to have as little deviation as possible, between the performance of typed abstractions (e.g `pmatch`, `pletFields`, `pfield`) and raw `pasConstr`. Currently `pmatch` in particular can generate inefficient code in certain cases.

We should also have benchmarks to track the progress of this.

**Comment by L-as** [comment]
This has already been fixed AFAICT.

**Comment by TotallyNotChase** [comment]
> This has already been fixed AFAICT.

Not quite, we still have some pretty significant inefficiencies in place. These are being tracked by the following benches-
* `types:data:deconstruction:matching:*:sumtype(ignore-fields)`
* `types:data:deconstruction:matching:*:sumtype(exhaustive)(ignore-fields)`
* `types:data:deconstruction:combined:*:toValidatorHash`

I meant this issue to track our progress on *optimizing* the stuff.
